### DISCUSSION TITLE
3DNow! and EMMI

### DISCUSSION THREAD

**Discussion by Torinde** [ptitSeb/box86#952] [Ideas]
What do you think of emulating [3DNow!](https://en.wikipedia.org/wiki/3DNow!) instructions (in both or either of box64 and box86)?
- 3DNow! emulation code can be taken from:
  - FEX-Emu for [all 3DNow!, including Extended and the Geode specific instructions](https://github.com/FEX-Emu/FEX/pull/1597)
  - QEMU patch [3dnow.diff](http://base91.sourceforge.net/download/qemu/3dnow.diff.gz) 
  - Bochs patch [3dnow.cc](https://sourceforge.net/p/bochs/code/HEAD/tree/trunk/bochs/cpu/3dnow.cc)
  - [86box](https://github.com/86Box/86Box)

I checked my_cpuid.c and it seems box86 and 64 are not emulating PAE, NX bit, SSE4a?

And what about [Extended MMX (Cyrix EMMI)](https://en.wikipedia.org/wiki/List_of_discontinued_x86_instructions#Cyrix_EMMI_instructions) instructions?
- There is a [CPU flag in Linux](https://unix.stackexchange.com/questions/43539/what-do-the-flags-in-proc-cpuinfo-mean) for those: `cxmmx` Cyrix MMX extensions
- [Cyrix Extension to the Multimedia Instruction Set, Application Note 108](http://hackipedia.org/browse.cgi/Hardware/CPU/x86/Cyrix%206x86/Cyrix%20Extension%20to%20the%20Multimedia%20Instruction%20Set%2c%20Application%20Note%20108%2c%20Applies%20to%20the%20Cyrix%20MII%e2%84%a2%20CPU%20%281998-09-09%29.pdf) (some further docs and software are linked [here](https://github.com/PCBox/PCBox/issues/39#issuecomment-1147292376))
- Opcodes overlap with SSE and other subsequent instructions, so it has to be selectable/option?
- https://github.com/86Box/86Box/pull/5284

**Comment by ptitSeb** [comment]
box86 / box64 are for now emulating an Intel CPU. Emulating an AMD cpu is possible of course, but I'm not sure how usefull it can be. There are handfull of games using 3DNow! and even less using 3DNowExt. SSE4a is also AMD specific.

If done, it needs to be parametrable (so you choose AMD instead of default Intel).

It's not on my short term or medium term todo. but I'll accept a PR if someone else want to have a look at it.

**Reply by Torinde** [reply]
Clear, thanks!

And what about PAE and NX bit?

**Reply by ptitSeb** [reply]
PAE has purpose only on OS level, user space program don't have any use of this.

NX is Itanium only iirc. It's for OS level too anyway. No user space program needs or use that.

**Reply by Torinde** [reply]
x86 also has [NX bit](https://en.wikipedia.org/wiki/NX_bit#x86). Is it useful (for compatibility?) to report via CPUID that PAE and NX are supported?

**Reply by ptitSeb** [reply]
I don't know if it's usefull. I can export the bits easily, for sure. I haven't seen any software complaining because they were not there. If you have some example, I gladly add the bits.

**Reply by Torinde** [reply]
OK. Last question for today: I see a few of the x86-64-v3 sets aren't supported yet: F16C, BMI1/2. Is there a technical reason for that or it's just a matter of priority/time?

**Reply by ptitSeb** [reply]
Many programs considers that if you have BMI, you have AVX.

**Reply by Torinde** [reply]
NX bit - Microsoft x86-to-ARM emulator is "providing" it (per the Windows CPUID Dumps at http://instlatx64.atw.hu - PAE is not set, NX is set), so probably it's good to mimic that.
